packages/core/src/queries/log.ts

```diff
@@ -1,42 +1,64 @@
-import type { HookExecutionStats, Log } from '@logto/schemas';
-import { token, Logs } from '@logto/schemas';
+import {
+  token,
+  type hook,
+  Logs,
+  type HookExecutionStats,
+  type Log,
+  type interaction,
+  type LogKeyUnknown,
+} from '@logto/schemas';
 import { conditionalSql, convertToIdentifiers } from '@logto/shared';
+import { conditionalArray } from '@silverhand/essentials';
 import { subDays } from 'date-fns';
-import type { CommonQueryMethods } from 'slonik';
 import { sql } from 'slonik';
+import type { CommonQueryMethods } from 'slonik';
 
 import { buildFindEntityByIdWithPool } from '#src/database/find-entity-by-id.js';
 import { buildInsertIntoWithPool } from '#src/database/insert-into.js';
 
 const { table, fields } = convertToIdentifiers(Logs);
 
+export type AllowedKeyPrefix = hook.Type | token.Type | interaction.Prefix | typeof LogKeyUnknown;
+
 type LogCondition = {
   logKey?: string;
-  applicationId?: string;
-  userId?: string;
-  hookId?: string;
+  payload?: { applicationId?: string; userId?: string; hookId?: string };
   startTimeExclusive?: number;
+  includeKeyPrefix?: AllowedKeyPrefix[];
+  excludeKeyPrefix?: AllowedKeyPrefix[];
 };
 
 const buildLogConditionSql = (logCondition: LogCondition) =>
-  conditionalSql(logCondition, ({ logKey, applicationId, userId, hookId, startTimeExclusive }) => {
-    const subConditions = [
-      conditionalSql(logKey, (logKey) => sql`${fields.key}=${logKey}`),
-      conditionalSql(userId, (userId) => sql`${fields.payload}->>'userId'=${userId}`),
-      conditionalSql(
-        applicationId,
-        (applicationId) => sql`${fields.payload}->>'applicationId'=${applicationId}`
-      ),
-      conditionalSql(hookId, (hookId) => sql`${fields.payload}->>'hookId'=${hookId}`),
-      conditionalSql(
-        startTimeExclusive,
-        (startTimeExclusive) =>
-          sql`${fields.createdAt} > to_timestamp(${startTimeExclusive}::double precision / 1000)`
-      ),
-    ].filter(({ sql }) => sql);
+  conditionalSql(
+    logCondition,
+    ({ logKey, payload, startTimeExclusive, includeKeyPrefix = [], excludeKeyPrefix = [] }) => {
+      const subConditions = [
+        ...conditionalArray(
+          includeKeyPrefix.length > 0 &&
+            includeKeyPrefix.map((prefix) => sql`${fields.key} like ${`${prefix}%`}`)
+        ),
+        ...conditionalArray(
+          excludeKeyPrefix.length > 0 &&
+            excludeKeyPrefix.map((prefix) => sql`${fields.key} not like ${`${prefix}%`}`)
+        ),
+        ...conditionalArray(
+          payload &&
+            Object.entries(payload).length > 0 &&
+            Object.entries(payload).map(([key, value]) =>
+              value ? sql`${fields.payload}->>'${key}'=${value}` : sql``
```

Comment:
map empty array should be fine? no need to check length > 0?

Reply:
removed the check on empty array